### PR TITLE
Use a context to propagate column-related Props, and remove `forceUpdate` usage

### PR DESCRIPTION
**Pull request by renchap**
The use of `forceUpdate` is prone to bugs and can not work with functional components.

I replaced it with a context providing the element to use to render the back button through a portal. I also added `multiColumn` to the context to avoid propagating it deep into the component tree, which will allow to remove it once we convert components to functional components.

The original issue this solved is discussed in #11630, #11605 and #11655. I tried to resize my window in single & advanced modes and it worked fine.

### CODE REVIEW

File: app/javascript/mastodon/components/column_back_button.tsx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+};
+
+export const ColumnBackButtonSlim: React.FC<{ onClick: OnClickCallback }> = ({
+  onClick,
+}) => {
+  const handleClick = useHandleClick(onClick);
+
+  return (
+    <div className='column-back-button--slim'>
+      {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events */}
```
Comment (renchap) [inline app/javascript/mastodon/components/column_back_button.tsx:54]: This is not ideal, and the correct fix here would be to convert the `<div>` to a proper `<button>`, but maybe there is a reason for it. to be a `<div>`?

File: app/javascript/mastodon/components/column_back_button.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+type OnClickCallback = () => void;
+
+function useHandleClick(onClick?: OnClickCallback) {
+  const history = useAppHistory();
+
+  return useCallback(() => {
+    if (onClick) {
+      onClick();
+    } else if (history.location.state.fromMastodon) {
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/components/column_back_button.tsx:20]: `history.location.state` can be `null` or `undefined`, e.g. it is `null` when not coming from a navigation event.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Looks way cleaner overall, but there are some unsafe assertions about the history state.

File: app/javascript/mastodon/components/column_back_button.tsx (hunk 1 above)
Comment (ClearlyClaire) [reply app/javascript/mastodon/components/column_back_button.tsx:54]: I can't think of a reason for it not to be a `button`. I am also unsure what the distinction for the “slim” is tbh.

Comment (renchap) [reply app/javascript/mastodon/components/column_back_button.tsx:54]: This is more complex than what I though, and needs bigger changes than the scope of this PR. I will open a new one for this.

**Review by ClearlyClaire — APPROVED** [review]
I'd like a bit more testing before this gets merged but it looks fine to me.

**Pull request merged**